Grammar/python.gram

```diff
@@ -640,8 +640,21 @@ invalid_assignment:
         RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "only single target (not tuple) can be annotated") }
     | a=expression ':' expression ['=' annotated_rhs] {
         RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "illegal target for annotation") }
-    | a=expression ('=' | augassign) (yield_expr | star_expressions) {
-        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "cannot assign to %s", _PyPegen_get_expr_name(a)) }
+    | a=star_expressions '=' (yield_expr | star_expressions) {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION( 
+            _PyPegen_get_invalid_target(a), 
+            "cannot assign to %s", _PyPegen_get_expr_name(_PyPegen_get_invalid_target(a))
+        )}
+    | a=expression augassign (yield_expr | star_expressions) {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION( 
+            _PyPegen_get_invalid_target(a), 
+            "cannot assign to %s", _PyPegen_get_expr_name(_PyPegen_get_invalid_target(a))
+        )}
```

Comment:
This probably segfaults, if the expression is a container that only contains valid targets, no?

Also, while the old parser indeed did output `cannot assign to function call` for `(a(), b) += 1, 2` I fear that that might be a misleading error message, since the tuple is itself an invalid augassign target. Wouldn't something like `cannot augassign to tuple` be a better error message here? Would there be backwards-compatibility problems if we were to do this?

Reply:
> This probably segfaults, if the expression is a container that only contains valid targets, no?

Yeah, I was dealing with this once I rebased....

> I fear that that might be a misleading error message, since the tuple is itself an invalid augassign target.

Yeah, we have in the old parser different error messages for when the target is a tuple/list of names and when there are illegal things inside. Not sure what is the best way to proceed. If we want to support both error messages we may need another function.

Reply:
> when the target is a tuple/list of names and when there are illegal things inside

Yeah, what I'm saying is that it shouldn't matter if there are valid or invalid things inside, since the whole tuple is itself invalid. IMHO we should just delete this alternative and let everything fall through to the other one:

```
| a=star_expressions augassign (yield_expr | star_expressions) {
        RAISE_SYNTAX_ERROR_KNOWN_LOCATION( 
            a, 
            "illegal expression for augmented assignment"
        )}
```

And maybe also work on the error message here to be a bit more descriptive.

Reply:
But if we do this, this error would be worse:

>>> f() += 3
  File "<stdin>", line 1
SyntaxError: cannot assign to function call

will now be

>>> f() += 3
  File "<stdin>", line 1
SyntaxError: illegal expression for augmented assignment

Reply:
In 896135f I have implemented a version that works like old parser.

Reply:
Right!

> And maybe also work on the error message here to be a bit more descriptive.

That's why I said this above. Maybe something like this?

```
| a=star_expressions augassign (yield_expr | star_expressions) {
        RAISE_SYNTAX_ERROR_KNOWN_LOCATION( 
            a, 
            "%s is an illegal expression for augmented assignment",
            _PyPegen_get_expr_type(a)
        )}
```

This would give us:

```
➜  cpython git:(pr/20076) ✗ ./python        
Python 3.9.0a6+ (heads/pr/20076-dirty:b596bbaa83, May 15 2020, 02:39:41) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> f() += 1
  File "<stdin>", line 1
    f() += 1
    ^
SyntaxError: function call is an illegal expression for augmented assignment
>>> (a, b) += 1
  File "<stdin>", line 1
    (a, b) += 1
    ^
SyntaxError: tuple is an illegal expression for augmented assignment
>>> (a(), b) += 1
  File "<stdin>", line 1
    (a(), b) += 1
    ^
SyntaxError: tuple is an illegal expression for augmented assignment
```

Reply:
> That's why I said this above. Maybe something like this?

Oh, gotcha! Apologies, this hour is not the best for me :worried:

Reply:
We would need to change also the old parser to match or somehow skip those cases for the old parser :S

Reply:
No worries at all. Is this okay though? Could a change like this be breaking code and if it does, should we worry about it?

Reply:
> No worries at all. Is this okay though? Could a change like this be breaking code and if it does, should we worry about it? What's the policy?

As long as the exception in the same I don't think there is a problem on changing the message. For instance, we normally improve error messages here and there between versions without any problem. The only thing that bothers me is that either we change this in the old parser as well or we deactivate the entire test because deactivating case by case is too much noise IMHO.

What do you think is best?

Reply:
I think changing the old parser would be best here. It's a very small change in `ast.c`.

Reply:
Done in c74ecce